apis/v1/policySnapshot_types.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package v1
+
+import (
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope="Cluster",shortName=pss,categories={fleet-workload}
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:JSONPath=`.metadata.generation`,name="Gen",type=string
+// +kubebuilder:printcolumn:JSONPath=`.metadata.creationTimestamp`,name="Age",type=date
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PolicySnapShot is used to store a snapshot of placement policy
+// It is immutable.
+type PolicySnapShot struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of PolicySnapShot.
+	// +required
+	Spec PolicySnapShotSpec `json:"spec"`
+
+	// The observed status of PolicySnapShot.
+	// +optional
+	Status PolicySnapShotStatus `json:"status,omitempty"`
+}
+
+// PolicySnapShotSpec defines the desired state of PolicySnapShot.
+type PolicySnapShotSpec struct {
+	// Policy defines how to select member clusters to place the selected resources.
+	// If unspecified, all the joined member clusters are selected.
+	// +optional
+	Policy *PlacementPolicy `json:"policy,omitempty"`
+
+	// IsLatest indicates if the policy is the latest or not.
+	// +required
+	IsLatest bool `json:"isLatest"`
```

Comment:
To reliably calculate the currently active policy snapshot, we would need ranks (unless we use sort by time, which is less reliable) + an isActive mark. Should we use labels for both marks?

Reply:
yes, we need label. I used a field to make it part of the schema. I wonder if we will lose the ability to do schema enforcement if we  use a label to replace a field? I know we can do validation webhook, is that the same? I am fine either way.